### PR TITLE
chore(artifacts): enable registry shorthand for more APIs

### PR DESCRIPTION
**Pull request by estellazx**
Description
-----------
https://wandb.atlassian.net/browse/WB-21583
For all paths that can potentially fetch Registry artifacts and projects. Add in code to fetch org entity for users so users can "shorthand" that path without passing an org entity in, or use the org display name

Apis changed in this PR:
- `api.project(name="wandb-registry-*")`
- `api.artifact_types()`
- `api.artifact_type()`
- `api.artifact_collections()`
- `api.artifact_collection()`
- `api.artifact_versions()` # Must include
- `api.artifacts()` # Must include

APIs not supported currently:
- `api.reports()`
- `api.job()/job_list` etc. anything to do with jobs
- `run_queue/queued_run`
- `api.sweep()`
- `api.runs()` # creation of runs in org entities is blocked
- `api.run()` # creation of runs in org entities is blocked
- `api.projects(entity="something")` # no way for us to know if you are trying to fetch registry projects

Testing
-------
Lots of manual testing. Had a team bug bash for it.
Made sure old code didn't break and new test worked:
```
api = wandb.Api(overrides={"entity": "team-entity"}) # switched between team and personal entity
p = api.project(name="wandb-registry-estella-test-registry")

arts = api.artifacts(name="wandb-registry-estella-test-registry/shorthand", type_name="raw_dataset)
at = api.artifact_type(type_name="raw_dataset", project="wandb-registry-estella-test-registry")
ats = api.artifact_types(project="wandb-registry-estella-test-registry")

acs = api.artifact_collections(
    project_name="wandb-registry-estella-test-registry", type_name="raw_dataset"
)
ac = api.artifact_collection(
    name="wandb-registry-estella-test-registry/shorthand",
    type_name="raw_dataset",
)

arts = api.artifacts(
    name="wandb-registry-estella-test-registry/shorthand", type_name="raw_dataset"
)
```

### CODE REVIEW

File: wandb/apis/public/api.py [hunk 1]
```diff
@@ -1041,7 +1049,12 @@ def artifact_types(self, project: Optional[str] = None) -> "public.ArtifactTypes
         Returns:
             An iterable `ArtifactTypes` object.
         """
-        entity, project = self._parse_project_path(project)
+        project_path = project
+        entity, project = self._parse_project_path(project_path)
+        # If its an Registry artifact, the entity is an org instead
```
Comment (ibindlish) [inline wandb/apis/public/api.py:1054]: **(nit)**
```suggestion
        # If its a Registry project, the entity is considered to be an org instead
```

File: wandb/apis/public/api.py [hunk 2]
```diff
@@ -1076,6 +1094,10 @@ def artifact_collections(
             An iterable `ArtifactCollections` object.
         """
         entity, project = self._parse_project_path(project_name)
+        # If iterating through Registry project, the entity is an org instead
```
Comment (ibindlish) [inline wandb/apis/public/api.py:1097]: **(nit)**
```suggestion
        # If iterating through Registry project, the entity is considered to be an org instead
```

File: wandb/apis/public/api.py [hunk 3]
```diff
@@ -1094,6 +1116,10 @@ def artifact_collection(
             An `ArtifactCollection` object.
         """
         entity, project, collection_name = self._parse_artifact_path(name)
+        # If its an Registry artifact, the entity the project belongs to is an org instead
```
Comment (ibindlish) [inline wandb/apis/public/api.py:1119]: **(nit)**
```suggestion
        # If its an Registry artifact, the entity is considered to be an org instead
```

File: wandb/apis/public/api.py [hunk 4]
```diff
@@ -1131,6 +1157,10 @@ def artifacts(
             An iterable `Artifacts` object.
         """
         entity, project, collection_name = self._parse_artifact_path(name)
+        # If its an Registry project, the entity is an org instead
```
Comment (ibindlish) [inline wandb/apis/public/api.py:1160]: **(nit)**
```suggestion
        # If its an Registry project, the entity is considered to be an org instead
```

File: wandb/apis/public/utils.py [hunk 5]
```diff
@@ -32,3 +34,20 @@ def parse_s3_url_to_s3_uri(url) -> str:
...
     return s3_uri
+
+
+def parse_org_from_registry_path(entity: str, project: str, path: str) -> str:
+    """Parse the org from a registry path.
+
+    Args:
+        entity (str): The entity name.
+        project (str): The project name.
+        path (str): The path to parse.
```
Comment (ibindlish) [inline wandb/apis/public/utils.py:45]: **(suggestion)** Nice! I think you could mention that the path here could be a project path or an artifact path

**Review by ibindlish — APPROVED** [review]
mostly lgtm! Just added a couple of nits to maintain consistency across the APIs and a minor docstring update suggestion. But deferring to you on those changes

File: wandb/apis/public/utils.py [hunk 6]
```diff
@@ -33,3 +35,20 @@ def parse_s3_url_to_s3_uri(url) -> str:
...
```
Comment (estellazx) [inline wandb/apis/public/utils.py:46]:
```suggestion
        path (str): The path to parse. Can be a project path <entity>/<project> or <project> or an 
        artifact path like <entity>/<project>/<artifact> or <project>/<artifact> or <artifact>
```

File: wandb/apis/public/utils.py [hunk 7]
```diff
@@ -33,3 +36,33 @@ def parse_s3_url_to_s3_uri(url) -> str:
...
+        path_type (PathType): The type of path to parse.
+    """
+    parts = path.split("/")
+    expected_parts = 3 if path_type == PathType.ARTIFACT else 2
+
+    if len(parts) >= expected_parts:
+        org, project = parts[:2]
+        if is_artifact_registry_project(project):
+            return org
+    return ""
```
Comment (kptkin) [inline wandb/apis/public/utils.py:68]: probably better to put it somewhere closer to artifacts it is literally only used for artifacts

**Review by kptkin — APPROVED** [review]
approve for stamp, didn't review the artifact changes

**Pull request merged**